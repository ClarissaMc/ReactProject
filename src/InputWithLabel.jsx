import * as React from 'react';

const InputWithLabel = ({ id, value, type = 'text', onInputChange, isFocused, children, }) => (
    <>
      <label htmlFor={id} className='label'>{children}</label>
      &nbsp;
      <input
        id={id}
        type={type}
        value={value}
        autoFocus={isFocused}
        onChange={onInputChange}
        className='input'
      />
    </>
);

export { InputWithLabel };