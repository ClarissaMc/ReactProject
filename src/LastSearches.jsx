import * as React from 'react';

const LastSearches = ({ lastSearches, onLastSearch }) => (
    <>
        {lastSearches.map((searchTerm, index) => (
            <button
                key={searchTerm + index}
                type="button"
                onClick={() => onLastSearch(searchTerm)}
            >
                {searchTerm}
            </button>
        ))}
    </>
);

export { LastSearches };