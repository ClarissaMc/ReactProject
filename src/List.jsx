import * as React from 'react';
import { sortBy } from 'lodash';

import Check from './check.svg?react';

const SORTS = {
    NONE: (list) => list,
    TITLE: (list) => sortBy(list, 'title'),
    AUTHOR: (list) => sortBy(list, 'author'),
    COMMENT: (list) => sortBy(list, 'num_comments').reverse(),
    POINT: (list) => sortBy(list, 'points').reverse(),
};

const List = React.memo(
    ({ list, onRemoveItem }) => {
    const [sort, setSort] = React.useState('NONE');

    const handleSort = (sortKey) => {
        const sortButtons = document.getElementsByClassName('button_sort');
        for (let i = 0; i < sortButtons.length; i++) {
            if (sortButtons[i].classList.contains(sortKey))
                sortButtons[i].classList.add('active_sort');
            else if (sortButtons[i].classList.contains('active_sort'))
                sortButtons[i].classList.remove('active_sort');
        }
        setSort(sortKey);
    };

    const sortFunction = SORTS[sort];
    const sortedList = sortFunction(list);

    return (
        <ul>
            <li style={{ display: 'flex' }}>
                <span style={{ width: '40%' }}>
                    <button 
                        type="button" 
                        className='button_sort TITLE' 
                        onClick={() => handleSort('TITLE')}
                    >
                        Title
                    </button>
                </span>
                <span style={{ width: '30%' }}>
                    <button
                        type="button"
                        className='button_sort AUTHOR'
                        onClick={() => handleSort('AUTHOR')}
                    >
                        Author
                    </button>
                </span>
                <span style={{ width: '10%' }}>
                    <button
                        type="button"
                        className='button_sort COMMENT'
                        onClick={() => handleSort('COMMENT')}
                    >
                        Comments
                    </button>
                </span>
                <span style={{ width: '10%' }}>
                    <button
                        type="button"
                        className='button_sort POINT'
                        onClick={() => handleSort('POINT')}
                    >
                        Points
                    </button>
                </span>
                <span style={{ width: '10%' }}>Actions</span>
            </li>

            {sortedList.map((item) => (
            <Item 
                key={item.objectID}
                item={item}
                onRemoveItem={onRemoveItem} />
            ))}
        </ul>
    );
    }
);
  
const Item = React.memo(
    ({ item, onRemoveItem }) => (
    <li className='item' style={{ display: 'flex' }}>
    <span style={{ width: '40%' }}>
        <a href={item.url}>{item.title}</a>
    </span>
    <span style={{ width: '30%' }}>{item.author}</span>
    <span style={{ width: '10%' }}>{item.num_comments}</span>
    <span style={{ width: '10%' }}>{item.points}</span>
    <span style={{ width: '10%' }}>
        <button 
            type="button"
            onClick={() => onRemoveItem(item)}
            className="button button_small"
        >
        <Check height="18px" width="18px"/>
        </button>
    </span>
    </li>
    )
);

export { List, Item };