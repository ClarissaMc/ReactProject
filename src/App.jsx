import * as React from 'react';
import axios from 'axios';

import { SearchForm } from './SearchForm';
import { List } from './List';

import './App.css';

const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';

const REMOVE_STORY = 'REMOVE_STORY';
const STORIES_FETCH_INIT = 'STORIES_FETCH_INIT';
const STORIES_FETCH_SUCCESS = 'STORIES_FETCH_SUCCESS';
const STORIES_FETCH_FAILURE = 'STORIES_FETCH_FAILURE';

const storiesReducer = (state, action) => {
  switch (action.type) {
    case STORIES_FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: false,
      };
    case STORIES_FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload,
      };
    case STORIES_FETCH_FAILURE:
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case REMOVE_STORY:
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();
  }
};

const useStorageState = (key, initialState) => {
  const isMounted = React.useRef(false);  // prevents initial re-render of component

  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initialState
  );

  React.useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = true;
    } else {
      localStorage.setItem(key, value);
    }
  }, [value, key]);

  return [value, setValue];
};

const App = () => {
  
  const [searchTerm, setSearchTerm] = useStorageState('search', 'React');

  const [url, setUrl] = React.useState(
    `${API_ENDPOINT}${searchTerm}`
  );

  const [stories, dispatchStories] = React.useReducer(
    storiesReducer,
    { data: [], isLoading: false, isError: false }  // this object is the state, which is named stories
  );

  const handleFetchStories = React.useCallback(async () => {
    dispatchStories({ type: STORIES_FETCH_INIT });

    try {
      const result = await axios.get(url);

      dispatchStories({
        type: STORIES_FETCH_SUCCESS,
        payload: result.data.hits,
      });
    } catch {
      dispatchStories({ type: STORIES_FETCH_FAILURE });
    }
  }, [url]);

  React.useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleRemoveStory = React.useCallback((item) => {
    dispatchStories({
      type: REMOVE_STORY,
      payload: item,
    });
  }, []);

  const handleSearchInput = React.useCallback((event) => {
    setSearchTerm(event.target.value);
  }, []);

  const handleSearchSubmit = (event) => {
    setUrl(`${API_ENDPOINT}${searchTerm}`);
    event.preventDefault();
  };

  return (
    <div className="container">
      <h1 className="headline-primary">My Hacker Stories</h1>

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      {stories.isError && <p>Something went wrong...</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
        <List
          list={stories.data}
          onRemoveItem={handleRemoveStory}
        />
      )}
    </div>
  );
};

export default App

export { storiesReducer, SearchForm, List };