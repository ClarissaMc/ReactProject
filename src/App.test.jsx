import { describe, it, expect, vi } from 'vitest';
import {
    render,
    screen,
    fireEvent,
    waitFor,
} from '@testing-library/react';

import App, {
    storiesReducer,
    List,
    SearchForm,
} from './App';

import { Item } from './List';
import { InputWithLabel } from './InputWithLabel';
import { LastSearches } from './LastSearches';

import axios from 'axios';

vi.mock('axios');

const storyOne = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
};

const storyTwo = {
    title: 'Redux',
    url: 'https://redux.js.org',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
};

const stories = [storyOne, storyTwo];

// --------------------------- Unit Tests ----------------------------------- //
describe('storiesReducer', () => {
    it('removes a story from all stories', () => {
        const action = { type: 'REMOVE_STORY', payload: storyOne };
        const state = { data: stories, isLoading: false, isError: false };

        const newState = storiesReducer(state, action);

        const expectedState = {
            data: [storyTwo],
            isLoading: false,
            isError: false,
        };

        expect(newState).toStrictEqual(expectedState);
    });

    it('initializes stories fetching', () => {
        const action = { type: 'STORIES_FETCH_INIT' };
        const state = { data: [], isLoading: false, isError: false };

        const newState = storiesReducer(state, action);

        const expectedState = {
            data: [],
            isLoading: true,
            isError: false,
        };

        expect(newState).toStrictEqual(expectedState);
    });

    it('successfully fetches stories', () => {
        const action = { type: 'STORIES_FETCH_SUCCESS', payload: stories };
        const state = { data: [], isLoading: true, isError: false };

        const newState = storiesReducer(state, action);

        const expectedState = {
            data: stories,
            isLoading: false,
            isError: false,
        };

        expect(newState).toStrictEqual(expectedState);
    });

    it('fails to fetch stories', () => {
        const action = { type: 'STORIES_FETCH_FAILURE' };
        const state = { data: [], isLoading: true, isError: false };

        const newState = storiesReducer(state, action);

        const expectedState = {
            data: [],
            isLoading: false,
            isError: true,
        };

        expect(newState).toStrictEqual(expectedState);
    });
});

describe('Item', () => {
    it('renders all properties', () => {
        render(<Item item={storyOne}/>);

        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.getByText('React')).toHaveAttribute(
            'href',
            'https://reactjs.org/'
        );
    });

    it('renders a clickable dismiss button', () => {
        render(<Item item={storyOne}/>);
        expect(screen.getByRole('button')).toBeInTheDocument();
    });

    it('clicking the dismiss button calls the callback handler', () => {
        const handleRemoveItem = vi.fn();

        render(<Item item={storyOne} onRemoveItem={handleRemoveItem}/>);

        fireEvent.click(screen.getByRole('button'));

        expect(handleRemoveItem).toHaveBeenCalledTimes(1);
        expect(handleRemoveItem).toHaveBeenCalledWith({...storyOne});
    });

    it('renders snapshot', () => {
        const { container } = render(<Item item={storyOne} onRemoveItem={vi.fn()}/>);
        expect(container.firstChild).toMatchSnapshot();
    });
});

describe('SearchForm', () => {
    const searchFormProps = {
        searchTerm: 'React',
        onSearchInput: vi.fn(),
        onSearchSubmit: vi.fn(),
    };

    it('renders the input field with its value', () => {
        render(<SearchForm {...searchFormProps}/>);
        expect(screen.getByDisplayValue('React')).toBeInTheDocument();
    });

    it('renders the correct label', () => {
        render(<SearchForm {...searchFormProps}/>);
        expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
    });

    it('calls onSearchInput on input field change', () => {
        render(<SearchForm {...searchFormProps}/>);

        fireEvent.change(screen.getByDisplayValue('React'), {
            target: { value: 'Redux' },
        });

        expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
    });

    it('calls onSearchSubmit on button submit click', () => {
        render(<SearchForm {...searchFormProps}/>);
        
        fireEvent.submit(screen.getByRole('button'));

        expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
    });

    it('renders snapshot', () => {
        const { container } = render(<SearchForm {...searchFormProps}/>);
        expect(container.firstChild).toMatchSnapshot();
    });
});

describe('InputWithLabel', () => {
    const inputWithLabelProps = {
        id: 'search',
        value: 'React',
        onInputChange: vi.fn(),
        children: <strong>Search:</strong>,
    };

    it('renders the input field with its value', () => {
        render(<InputWithLabel {...inputWithLabelProps}/>);
        expect(screen.getByDisplayValue('React')).toBeInTheDocument();
    });

    it('renders the correct label', () => {
        render(<InputWithLabel {...inputWithLabelProps}/>);
        expect(screen.getByLabelText(/Search/)).toBeInTheDocument();
    });

    it('calls onInputSearch on input field change', () => {
        render(<InputWithLabel {...inputWithLabelProps}/>);
        
        fireEvent.change(screen.getByDisplayValue('React'), {
            target: { value: 'Redux' },
        });

        expect(inputWithLabelProps.onInputChange).toHaveBeenCalledTimes(1);
    });

    it('renders snapshot', () => {
        const { container } = render(<InputWithLabel {...inputWithLabelProps}/>);
        expect(container.firstChild).toMatchSnapshot();
    });
});

describe('List', () => {
    const listProps = {
        list: stories,
        onRemoveItem: vi.fn(),
    };

    it('renders the stories', () => {
        render(<List {...listProps}/>);
        expect(screen.getByRole('list')).toBeInTheDocument();
        expect(screen.getByText('React')).toBeInTheDocument();
        expect(screen.getByText('Redux')).toBeInTheDocument();
    });

    it('renders the sort buttons', () => {
        render(<List {...listProps}/>);
        expect(screen.getAllByRole('button').length).toBe(6);
        expect(screen.getByText('Title')).toBeInTheDocument();
        expect(screen.getByText('Author')).toBeInTheDocument();
        expect(screen.getByText('Comments')).toBeInTheDocument();
        expect(screen.getByText('Points')).toBeInTheDocument();
        expect(screen.getByText('Actions')).toBeInTheDocument();
    });

    it('fires sort buttons', () => {
        const {container } = render(<List {...listProps}/>);

        expect(container.getElementsByClassName('active_sort').length).toBe(0);

        fireEvent.click(screen.getAllByRole('button')[0]);  // Title

        expect(container.getElementsByClassName('active_sort').length).toBe(1);
        expect(container.getElementsByClassName('active_sort')[0].classList.contains('TITLE')).toBeTruthy();

        fireEvent.click(screen.getAllByRole('button')[1]);  // Author

        expect(container.getElementsByClassName('active_sort').length).toBe(1);
        expect(container.getElementsByClassName('active_sort')[0].classList.contains('AUTHOR')).toBeTruthy();
    });

    it('renders sort arrows', () => {
        render(<List {...listProps}/>);

        expect(screen.queryByTitle('down-arrow')).toBeNull();
        expect(screen.queryByTitle('up-arrow')).toBeNull();

        fireEvent.click(screen.getAllByRole('button')[0]);  // Title

        expect(screen.queryByTitle('down-arrow')).toBeInTheDocument();
        expect(screen.queryByTitle('up-arrow')).toBeNull();

        fireEvent.click(screen.getAllByRole('button')[0]);  // Title Reversed

        expect(screen.queryByTitle('down-arrow')).toBeNull();
        expect(screen.queryByTitle('up-arrow')).toBeInTheDocument();

        fireEvent.click(screen.getAllByRole('button')[3]);  // Comment

        expect(screen.queryByTitle('down-arrow')).toBeInTheDocument();
        expect(screen.queryByTitle('up-arrow')).toBeNull();

        fireEvent.click(screen.getAllByRole('button')[3]);  // Comment Reversed

        expect(screen.queryByTitle('down-arrow')).toBeNull();
        expect(screen.queryByTitle('up-arrow')).toBeInTheDocument();
    });

    it('renders snapshot', () => {
        const { container } = render(<List {...listProps}/>);
        expect(container.firstChild).toMatchSnapshot();
    });
});

describe('LastSearches', () => {
    const lastSearches = [
        'React',
        'Redux',
        'Java',
        'Python',
        'C',
    ]
    const lastSearchesProps = { lastSearches: lastSearches, onLastSearch: vi.fn() };

    it('renders the buttons', () => {
        render(<LastSearches {...lastSearchesProps}/>);
        expect(screen.getAllByRole('button').length).toBe(5);
        expect(screen.getByText('React')).toBeInTheDocument();
        expect(screen.getByText('Redux')).toBeInTheDocument();
        expect(screen.getByText('Java')).toBeInTheDocument();
        expect(screen.getByText('Python')).toBeInTheDocument();
        expect(screen.getByText('C')).toBeInTheDocument();
    });

    it('calls onLastSearch when button clicked', () => {
        render(<LastSearches {...lastSearchesProps}/>);
        fireEvent.click(screen.getByText('Java'));
        expect(lastSearchesProps.onLastSearch).toHaveBeenCalledTimes(1);
    });

    it('renders snapshot', () => {
        const { container } = render(<LastSearches {...lastSearchesProps}/>);
        expect(container.firstChild).toMatchSnapshot();
    });
});
// ------------------------- End Unit Tests --------------------------------- //

// ----------------------- Integration Tests -------------------------------- //
describe('App', () => {
    it('succeeds fetching data', async () => {
        const promise = Promise.resolve({
            data: {
                hits: stories,
            },
        });

        axios.get.mockImplementationOnce(() => promise);

        render(<App/>);

        expect(screen.queryByText(/Loading/)).toBeInTheDocument();

        await waitFor(async () => await promise);

        expect(screen.queryByText(/Loading/)).toBeNull();

        expect(screen.getByText('React')).toBeInTheDocument();
        expect(screen.getByText('Redux')).toBeInTheDocument();
    });

    it('fails fetching data', async () => {
        const promise = Promise.reject();
        axios.get.mockImplementationOnce(() => promise);

        render(<App/>);

        expect(screen.getByText(/Loading/)).toBeInTheDocument();

        try {
            await waitFor(async () => await promise);
        } catch (error) {
            expect(screen.queryByText(/Loading/)).toBeNull();
            expect(screen.queryByText(/went wrong/)).toBeInTheDocument();
        }
    });

    it('removes a story', async () => {
        const promise = Promise.resolve({
            data: {
                hits: stories,
            },
        });

        axios.get.mockImplementationOnce(() => promise);

        render(<App/>);

        await waitFor(async () => await promise);

        expect(screen.getAllByRole('button').length).toBe(7);
        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();

        fireEvent.click(screen.getAllByRole('button')[5]);

        expect(screen.getAllByRole('button').length).toBe(6);
        expect(screen.queryByText('Jordan Walke')).toBeNull();
    });

    it('searches for specific stories', async () => {
        // tests the initial request and another request

        const reactPromise = Promise.resolve({
            data: {
                hits: stories,
            },
        });

        const anotherStory = {
            title: 'JavaScript',
            url: 'https://en.wikipedia.org/wiki/JavaScript',
            author: 'Brendan Eich',
            num_comments: 15,
            points: 10,
            objectID: 3,
        };

        const javascriptPromise = Promise.resolve({
            data: {
                hits: [anotherStory],
            },
        });

        axios.get.mockImplementation((url) => {
            if (url.includes('React')) {
                return reactPromise;
            }
            if (url.includes('JavaScript')) {
                return javascriptPromise;
            }

            throw Error();
        });

        // Initial Render

        render(<App/>);

        // First Data Fetching

        await waitFor(async () => await reactPromise);

        expect(screen.queryByDisplayValue('React')).toBeInTheDocument();
        expect(screen.queryByDisplayValue('JavaScript')).toBeNull();

        expect(screen.queryByText('Jordan Walke')).toBeInTheDocument();
        expect(
            screen.queryByText('Dan Abramov, Andrew Clark')
        ).toBeInTheDocument();
        expect(screen.queryByText('Brendan Eich')).toBeNull();

        // User Interaction -> Search

        fireEvent.change(screen.queryByDisplayValue('React'), {
            target: {
                value: 'JavaScript',
            },
        });

        expect(screen.queryByDisplayValue('React')).toBeNull();
        expect(
            screen.queryByDisplayValue('JavaScript')
        ).toBeInTheDocument();

        fireEvent.submit(screen.queryByText('Submit'));

        // Second Data Fetching

        await waitFor(async () => await javascriptPromise);

        expect(screen.queryByText('Jordan Walke')).toBeNull();
        expect(
            screen.queryByText('Dan Abramov, Andrew Clark')
        ).toBeNull();
        expect(screen.queryByText('Brendan Eich')).toBeInTheDocument();
    });

    it('sorts the list', async () => {
        const promise = Promise.resolve({
            data: {
                hits: stories,
            },
        });

        axios.get.mockImplementationOnce(() => promise);

        render(<App/>);

        await waitFor(async () => await promise);

        const reactElement = screen.getByText('React');
        const reduxElement = screen.getByText('Redux');

        expect(reactElement.compareDocumentPosition(reduxElement)).toBe(4);

        fireEvent.click(screen.getAllByRole('button')[2]);  // Author
        expect(reactElement.compareDocumentPosition(reduxElement)).toBe(2);
        fireEvent.click(screen.getAllByRole('button')[2]);  // Author Reverse
        expect(reactElement.compareDocumentPosition(reduxElement)).toBe(4);

        fireEvent.click(screen.getAllByRole('button')[1]);  // Title
        expect(reactElement.compareDocumentPosition(reduxElement)).toBe(4);
        fireEvent.click(screen.getAllByRole('button')[1]);  // Title Reverse
        expect(reactElement.compareDocumentPosition(reduxElement)).toBe(2);

        fireEvent.click(screen.getAllByRole('button')[3]);  // Comments
        expect(reactElement.compareDocumentPosition(reduxElement)).toBe(4);
        fireEvent.click(screen.getAllByRole('button')[3]);  // Comments Reverse
        expect(reactElement.compareDocumentPosition(reduxElement)).toBe(2);

        fireEvent.click(screen.getAllByRole('button')[4]);  // Points
        expect(reactElement.compareDocumentPosition(reduxElement)).toBe(2);
        fireEvent.click(screen.getAllByRole('button')[4]);  // Points Reverse
        expect(reactElement.compareDocumentPosition(reduxElement)).toBe(4);
    });

    it('updates InputWithLabel when LastSearches button clicked', async () => {
        const promise = Promise.resolve({
            data: {
                hits: stories,
            },
        });

        axios.get.mockImplementationOnce(() => promise);

        render(<App/>);

        await waitFor(async () => await promise);

        expect(screen.getByDisplayValue('JavaScript')).toBeInTheDocument();

        fireEvent.click(screen.getByText('Submit'));    // 'JavaScript' search
        fireEvent.change(screen.getByDisplayValue('JavaScript'), {
            target: { value: 'React' },
        });
        fireEvent.click(screen.getByText('Submit'));    // 'React' search

        expect(screen.getByDisplayValue('React')).toBeInTheDocument();
        expect(screen.getByText('JavaScript')).toBeInTheDocument();

        fireEvent.click(screen.getByText('JavaScript')); // 'JavaScript' search

        expect(screen.getByDisplayValue('JavaScript')).toBeInTheDocument();
    });

    it('adds new last search button', async () => {
        const promise = Promise.resolve({
            data: {
                hits: stories,
            },
        });

        axios.get.mockImplementationOnce(() => promise);

        render(<App/>);

        await waitFor(async () => await promise);

        expect(screen.getByDisplayValue('JavaScript')).toBeInTheDocument();

        fireEvent.click(screen.getByText('Submit'));    // 'JavaScript' search
        fireEvent.change(screen.getByDisplayValue('JavaScript'), {
            target: { value: 'React' },
        });
        fireEvent.click(screen.getByText('Submit'));    // 'React' search

        expect(screen.queryAllByText('JavaScript').length).toBe(1);
        expect(screen.queryAllByText('React').length).toBe(0);

        fireEvent.click(screen.getByText('Submit'));    // 'React' search

        expect(screen.queryAllByText('JavaScript').length).toBe(1);
        expect(screen.queryAllByText('React').length).toBe(0);

        fireEvent.change(screen.getByDisplayValue('React'), {
            target: { value: 'JavaScript' },
        });
        fireEvent.click(screen.getByText('Submit'));    // 'JavaScript' search

        expect(screen.queryAllByText('JavaScript').length).toBe(1);
        expect(screen.queryAllByText('React').length).toBe(1);

        fireEvent.change(screen.getByDisplayValue('JavaScript'), {
            target: { value: 'Python' },
        });
        fireEvent.click(screen.getByText('Submit'));    // 'Python' search

        expect(screen.queryAllByText('JavaScript').length).toBe(2);
        expect(screen.queryAllByText('React').length).toBe(2);  // last search & story
    });
});
// ----------------------- End Integration Tests ---------------------------- //