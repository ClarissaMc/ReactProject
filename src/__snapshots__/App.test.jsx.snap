// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputWithLabel > renders snapshot 1`] = `
<label
  class="label"
  for="search"
>
  <strong>
    Search:
  </strong>
</label>
`;

exports[`Item > renders snapshot 1`] = `
<li
  class="item"
>
  <span
    style="width: 40%;"
  >
    <a
      href="https://reactjs.org/"
    >
      React
    </a>
  </span>
  <span
    style="width: 30%;"
  >
    Jordan Walke
  </span>
  <span
    style="width: 10%;"
  >
    3
  </span>
  <span
    style="width: 10%;"
  >
    4
  </span>
  <span
    style="width: 10%;"
  >
    <button
      class="button button_small"
      type="button"
    >
      <svg
        height="18px"
        id="Capa_1"
        style="enable-background: new 0 0 297 297;"
        viewBox="0 0 297 297"
        width="18px"
        x="0px"
        xml:space="preserve"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
        y="0px"
      >
        <g>
          <path
            d="M113.636,272.638c-2.689,0-5.267-1.067-7.168-2.97L2.967,166.123c-3.956-3.957-3.956-10.371-0.001-14.329l54.673-54.703       c1.9-1.9,4.479-2.97,7.167-2.97c2.689,0,5.268,1.068,7.169,2.969l41.661,41.676L225.023,27.332c1.9-1.901,4.48-2.97,7.168-2.97l0,0       c2.688,0,5.268,1.068,7.167,2.97l54.675,54.701c3.956,3.957,3.956,10.372,0,14.328L120.803,269.668       C118.903,271.57,116.325,272.638,113.636,272.638z M24.463,158.958l89.173,89.209l158.9-158.97l-40.346-40.364L120.803,160.264       c-1.9,1.902-4.478,2.971-7.167,2.971c-2.688,0-5.267-1.068-7.168-2.97l-41.66-41.674L24.463,158.958z"
          />
        </g>
      </svg>
    </button>
  </span>
</li>
`;

exports[`List > renders snapshot 1`] = `
<ul>
  <li
    style="display: flex;"
  >
    <span
      style="width: 40%;"
    >
      Title
    </span>
    <span
      style="width: 30%;"
    >
      Author
    </span>
    <span
      style="width: 10%;"
    >
      Comments
    </span>
    <span
      style="width: 10%;"
    >
      Points
    </span>
    <span
      style="width: 10%;"
    >
      Actions
    </span>
  </li>
  <li
    class="item"
  >
    <span
      style="width: 40%;"
    >
      <a
        href="https://reactjs.org/"
      >
        React
      </a>
    </span>
    <span
      style="width: 30%;"
    >
      Jordan Walke
    </span>
    <span
      style="width: 10%;"
    >
      3
    </span>
    <span
      style="width: 10%;"
    >
      4
    </span>
    <span
      style="width: 10%;"
    >
      <button
        class="button button_small"
        type="button"
      >
        <svg
          height="18px"
          id="Capa_1"
          style="enable-background: new 0 0 297 297;"
          viewBox="0 0 297 297"
          width="18px"
          x="0px"
          xml:space="preserve"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          y="0px"
        >
          <g>
            <path
              d="M113.636,272.638c-2.689,0-5.267-1.067-7.168-2.97L2.967,166.123c-3.956-3.957-3.956-10.371-0.001-14.329l54.673-54.703       c1.9-1.9,4.479-2.97,7.167-2.97c2.689,0,5.268,1.068,7.169,2.969l41.661,41.676L225.023,27.332c1.9-1.901,4.48-2.97,7.168-2.97l0,0       c2.688,0,5.268,1.068,7.167,2.97l54.675,54.701c3.956,3.957,3.956,10.372,0,14.328L120.803,269.668       C118.903,271.57,116.325,272.638,113.636,272.638z M24.463,158.958l89.173,89.209l158.9-158.97l-40.346-40.364L120.803,160.264       c-1.9,1.902-4.478,2.971-7.167,2.971c-2.688,0-5.267-1.068-7.168-2.97l-41.66-41.674L24.463,158.958z"
            />
          </g>
        </svg>
      </button>
    </span>
  </li>
  <li
    class="item"
  >
    <span
      style="width: 40%;"
    >
      <a
        href="https://redux.js.org"
      >
        Redux
      </a>
    </span>
    <span
      style="width: 30%;"
    >
      Dan Abramov, Andrew Clark
    </span>
    <span
      style="width: 10%;"
    >
      2
    </span>
    <span
      style="width: 10%;"
    >
      5
    </span>
    <span
      style="width: 10%;"
    >
      <button
        class="button button_small"
        type="button"
      >
        <svg
          height="18px"
          id="Capa_1"
          style="enable-background: new 0 0 297 297;"
          viewBox="0 0 297 297"
          width="18px"
          x="0px"
          xml:space="preserve"
          xmlns="http://www.w3.org/2000/svg"
          xmlns:xlink="http://www.w3.org/1999/xlink"
          y="0px"
        >
          <g>
            <path
              d="M113.636,272.638c-2.689,0-5.267-1.067-7.168-2.97L2.967,166.123c-3.956-3.957-3.956-10.371-0.001-14.329l54.673-54.703       c1.9-1.9,4.479-2.97,7.167-2.97c2.689,0,5.268,1.068,7.169,2.969l41.661,41.676L225.023,27.332c1.9-1.901,4.48-2.97,7.168-2.97l0,0       c2.688,0,5.268,1.068,7.167,2.97l54.675,54.701c3.956,3.957,3.956,10.372,0,14.328L120.803,269.668       C118.903,271.57,116.325,272.638,113.636,272.638z M24.463,158.958l89.173,89.209l158.9-158.97l-40.346-40.364L120.803,160.264       c-1.9,1.902-4.478,2.971-7.167,2.971c-2.688,0-5.267-1.068-7.168-2.97l-41.66-41.674L24.463,158.958z"
            />
          </g>
        </svg>
      </button>
    </span>
  </li>
</ul>
`;

exports[`SearchForm > renders snapshot 1`] = `
<form
  class="search-form"
>
  <label
    class="label"
    for="search"
  >
    <strong>
      Search:
    </strong>
  </label>
  Â 
  <input
    class="input"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="button button_large"
    type="submit"
  >
    Submit
  </button>
</form>
`;
